src/MergeContext.js

```diff
@@ -499,58 +557,90 @@ class MergeContext {
+            if (requiredContexts.some(el => el.trim() === st.context.trim() && el.trim() !== Config.approvalContext()))
+                statusResult.addRequiredStatus(new StatusCheck(st));
+        }
+        return statusResult;
+    }
+
+    async _checkPRStatuses(logContext) {
+        const prStatus = await this._getPRStatuses();
+        this._log("status details: " + prStatus);
+        if (prStatus.succeeded())
+            return StepResult.Succeed();
+        // non-final or failed
+        this._log(logContext + " 'status' failed");
+        return StepResult.Fail();
+    }
+
+    // returns filled StatusResult object
+    async _getStagingStatuses() {
+        const combinedStagingStatuses = await GH.getStatuses(this._tagSha);
+        const genuineStatuses = combinedStagingStatuses.statuses.filter(st => !st.description.endsWith(Config.copiedDescriptionSuffix()));
+        assert(genuineStatuses.length <= Config.stagingChecks());
+        let statusResult = new StatusResult(Config.stagingChecks(), "Staging");
+        // all checks are 'required'
+        for (let st of genuineStatuses)
+            statusResult.addRequiredStatus(new StatusCheck(st));
+        return statusResult;
+    }
+
+    async _processStagingStatuses() {
+        const stagingStatus = await this._getStagingStatuses();
+        this._log("status details: " + stagingStatus);
+        if (stagingStatus.failed()) {
+            this._log("staging checks failed");
+            return await this._cleanupMergeFailed(false, this._labelFailedStagingChecks);
+        } else if (!stagingStatus.final()) {
+            if (!this._dryRun("setting M-wating-staging-checks label"))
+                await this._labelWaitingStagingChecks();
+            this._log("waiting for more staging checks completing");
+            return StepResult.Suspend();
         }
+        assert(stagingStatus.succeeded());
+        this._log("staging checks succeeded");
+        if (this._dryRun("finish processing"))
+            return StepResult.Suspend();
+        await this._supplyStagingWithPrRequired(stagingStatus);
+        return StepResult.Succeed();
+    }
+
+    // Creates PR-required status checks for staged commit (if possible).
+    // Staged commit needs all PR-required checks (configured on GitHub)
+    // so that GitHub could merge it into the protected base branch.
+    async _supplyStagingWithPrRequired(stagedStatuses) {
+        assert(stagedStatuses.succeeded());
 
-        return ret ? 'success' : 'failure';
+        const requiredPrContexts = await this._getRequiredContexts();
+        const prStatuses = await this._getPRStatuses();
+
+        for (let requiredContext of requiredPrContexts) {
+            if (!stagedStatuses.requiredStatuses.some(el => el.context.trim() === requiredContext.trim())) {
+                const requiredPrStatus = prStatuses.requiredStatuses.find(el => el.context.trim() === requiredContext.trim());
+                assert(requiredPrStatus);
```

Comment:
Let's also assert that `requiredPrStatus.description` does not end with copiedDescriptionSuffix().